tests/utils/test_python_manager.py

```diff
@@ -0,0 +1,116 @@
+    python = Python(executable=sys.executable)
+
+    assert python.executable == Path(sys.executable)
+    assert python.version == Version.parse(
+        ".".join([str(s) for s in sys.version_info[:3]])
+    )
+    assert python.patch_version == Version.parse(
+        ".".join([str(s) for s in sys.version_info[:3]])
+    )
+    assert python.minor_version == Version.parse(
+        ".".join([str(s) for s in sys.version_info[:2]])
+    )
+
+
+def test_python_get_system_python() -> None:
+    python = Python.get_system_python()
+
+    assert python.executable == Path(sys.executable)
+    assert python.version == Version.parse(
+        ".".join(str(v) for v in sys.version_info[:3])
+    )
+
+
+def test_python_get_preferred_default(config: Config) -> None:
+    python = Python.get_preferred_python(config)
+
+    assert python.executable == Path(sys.executable)
+    assert python.version == Version.parse(
+        ".".join(str(v) for v in sys.version_info[:3])
+    )
+
+
+def test_python_get_preferred_activated(config: Config, mocker: MockerFixture) -> None:
+    mocker.patch(
+        "subprocess.check_output",
+        side_effect=check_output_wrapper(Version.parse("3.7.1")),
+    )
+    config.config["virtualenvs"]["prefer-active-python"] = True
+    python = Python.get_preferred_python(config)
+
+    assert python.executable.as_posix().startswith("/usr/bin/python")
+    assert python.version == Version.parse("3.7.1")
+
+
+def test_python_get_preferred_activated_fallback(
+    config: Config, mocker: MockerFixture
+) -> None:
+    config.config["virtualenvs"]["prefer-active-python"] = True
+    with mocker.patch(
+        "subprocess.check_output",
+        side_effect=subprocess.CalledProcessError(1, "some command"),
+    ):
+        python = Python.get_preferred_python(config)
+
+    assert python.executable == Path(sys.executable)
+
+
+def test_fallback_on_detect_active_python(mocker: MockerFixture) -> None:
+    m = mocker.patch(
+        "subprocess.check_output",
+        side_effect=subprocess.CalledProcessError(1, "some command"),
+    )
+
+    active_python = Python._detect_active_python(NullIO())
+
+    assert active_python is None
+    assert m.call_count == 1
+
+
+@pytest.mark.skipif(sys.platform != "win32", reason="Windows only")
+def test_detect_active_python_with_bat(tmp_path: Path) -> None:
+    """On Windows pyenv uses batch files for python management."""
+    python_wrapper = tmp_path / "python.bat"
+    wrapped_python = Path(r"C:\SpecialPython\python.exe")
+    encoding = "locale" if sys.version_info >= (3, 10) else None
+    with python_wrapper.open("w", encoding=encoding) as f:
+        f.write(f"@echo {wrapped_python}")
+    os.environ["PATH"] = str(python_wrapper.parent) + os.pathsep + os.environ["PATH"]
+
+    active_python = Python._detect_active_python(NullIO())
+
+    assert active_python == wrapped_python
+
+
+def test_python_find_compatible(project_factory: ProjectFactory) -> None:
```

Comment:
Interesting finding :+1: I believe it's not a problem with the test. The test discovers a bug/suboptimal implementation. 

The logic within `get_compatible_python()` assumes that there as an executable with a version number in the name (e.g. `python3`, `python312`). The assumption is obviously not correct. We didn't run into any trouble until now, because this logic is only triggered if the "system" or "preferred" python isn't compatible. So an executable `python` was checked before.

We could add  `python` to  the list of executables that are tried. But where? At the beginning or at the end?

Reply:
> We could add python to the list of executables that are tried. But where? At the beginning or at the end?

I like adding it but do not have a strong opinion where to add it.

Reply:
I have added it to the end.

Reply:
I think this test has a high chance of flaking on random Windows systems until we have #2117 resolved (i.e. sometime after this PR) since a 'python.exe' in the path on Windows has a reasonable chance of being the Store-launcher version (which we would hopefully _reject_ if it was chosen...) or simply not existing at all.

We _could_ also try 'py.exe' after 'python.exe' which would give you _some_ PEP-514 version of Python, if possible. I just noticed on my machine that it picked up the Python 3.13.0rc1 experimental free-threaded build in preference to either the traditional 3.13.0rc1, or (released) 3.12, so yeah... it's a lexicographical version ordering. Surprise!

But I don't think we try 'py.exe' now, so I don't think that change would make sense in this refactoring anyway.

Is there value in marking this test "May fail due to system configuration on Windows, see #2117" or the like? (I am _assuming_ that the implementation for #2117 will work for both `py use env` and `get_compatible_python`...)

Reply:
Ok, but from what I understood, this PR is just refactoring, not bug fixing. If there are bugs that were present previously, we can take care of those in a separate PR.

Reply:
Yes, that's why my suggestion was to simply annotate the test, since it's a _newly distinguished function_ in this refactoring, and the new test for it has flake-potential in rational-but-uncontrolled environments, and a possibly-related existing bug report where we can track and resolve (or at least improve) that.

Reply:
Added a comment as suggested.

Reply:
Just looking at the comment, I think it might have conflated a few things:
* The Store-launcher that [is bundled in Windows was python.exe in the path](https://learn.microsoft.com/windows/python/faqs#why-does-running-python-exe-open-the-microsoft-store-). I expect that's rejected by logic that checks the version of the discovered Python binary, but would pop up the store UI in the meantime.
* A store-installed Python will put functioning python.exe and python3.exe in the path (apparently, I didn't test this), causing this test to work correctly, and giving the most POSIX-like experience in this respect.
* A python.org installer may or may not put python.exe in the path (not sure of the default off-hand, I always have it off personally), but definitely won't put python3.exe in the path.
* A python.org installer would normally install a py.exe which behaves link the normal /usr/bin/python symlink on POSIX machines, as far as `py -c` goes, anyway.

So I think
```python
    # Note: This test may fail on Windows systems using Python from the Microsoft Store,
    # as the executable is named `py.exe`, which is not currently recognized by
    # Python.get_compatible_python. This issue will be resolved in #2117.
```

maybe should be

```python
    # Note: This test may fail on Windows systems using Python from the python.org installer,
    # as the included `python.exe` may not be on the PATH, and hence not seen by
    # Python.get_compatible_python. This issue will be resolved in #2117.
    # Note: On Windows systems with nothing else putting a `python.exe` in the PATH,
    # this test may open the Microsoft Store, see
    # https://learn.microsoft.com/windows/python/faqs#why-does-running-python-exe-open-the-microsoft-store-
```

(I added the second note because #2117 won't address that, so it'll have to be left behind when #2117 is done and the first note is removed. There's basically nothing we can do for that case except special case that exact executable, perhaps by full path in `_full_python_path`, and that's something we can fuss about with later if it's interesting.)

That said, looking at the rest of the comment, if we can guarantee that the Poetry-running Python executable will be selected in this case, maybe the comment could be more-succinctly summarised to just that, or even asserted somehow? Although with a quick look, I don't see how this happens? `get_compatible_python` doesn't appear to try the result of `get_system_python` before it goes searching the PATH with `_full_python_path`, so if our currently-executing python binary isn't in the PATH, it won't be found, I think.